### PR TITLE
mzbuild: support configuring image registry and prefix

### PR DESCRIPTION
**Pull request by benesch**
To support use from the cloud repo, which wants to publish their mzbuild images to a ECR registry that's shared among many personal stacks.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: misc/python/materialize/mzbuild.py [hunk 1]
```diff
@@ -465,13 +472,16 @@ def sync_description(self) -> None:
...
+                self.docker_name(),
             ]
         )
 
-    def docker_name(self, tag: str) -> str:
+    def docker_name(self, tag: str | None = None) -> str:
         """Return the name of the image on Docker Hub at the given tag."""
-        return f"materialize/{self.name}:{tag}"
+        name = f"{self.rd.image_registry}/{self.rd.image_prefix}{self.name}"
+        if tag:
```
Comment (def-) [inline misc/python/materialize/mzbuild.py:482]:
```suggestion
        if tag is not None:
```
But an empty `""` tag probably doesn't make much sense anyway.

File: misc/python/materialize/mzbuild.py [hunk 2]
```diff
@@ -960,7 +992,7 @@ def publish_multiarch_images(
     for images in zip(*dependency_sets):
         names = set(image.image.name for image in images)
         assert len(names) == 1, "dependency sets did not contain identical images"
```
Comment (def-) [inline misc/python/materialize/mzbuild.py:994]:
```suggestion
        assert len(images) == 1, "dependency sets did not contain identical images"
```

**Review by def- — APPROVED** [review]
Just minor suggestions.

File: misc/python/materialize/mzbuild.py (hunk 2 above)
Comment (alex-hunt-materialize) [reply misc/python/materialize/mzbuild.py:994]: Do we actually want this change? Could there be duplicate images brought in as transitive dependencies? If so, the current way of creating a set of their names dedupes them.

**Review by alex-hunt-materialize — APPROVED** [review]

File: misc/python/materialize/mzbuild.py (hunk 2 above)
Comment (def-) [reply misc/python/materialize/mzbuild.py:994]: You're right, I was wrong!

Comment (benesch) [reply misc/python/materialize/mzbuild.py:994]: This is for publishing multiarch images, so it's expected that there are exactly two images with the same name, one for x86 and one for ARM. The code is structured such that it generalizes to more than two architectures, too!

**Pull request merged**